### ISSUE TITLE
[Documentation] Document how to @beartype JAX dataclasses with Equinox

### ISSUE THREAD

**Issue by murphyk** [beartype/beartype#185]
This works without type checking, but fails when I add the `@typechecker` decorator
```
import jax.numpy as jnp
from jax import vmap
from typing import NamedTuple, Optional, Union
from beartype import beartype as typechecker
#from typeguard import typechecked as typechecker


@typechecker
class MyParams(NamedTuple):
    mean: float

theta = MyParams(42.0)

def f(params, y):
    return params.mean + y

data = jnp.ones((10,2))
foo = vmap(f, (None,0))(theta, data)
print(foo)
```
It raises the error
```
@beartyped namedtuple_MyParams.MyParams.__new__()
 parameter mean=<object object at 0x7fb770c3a840> violates type hint <class 'float'>, 
as object <object object at 0x7fb770c3a840> not instance of float.
```
If I use typeguard i get the error
```
type of argument "mean" must be either float or int; got object instead
```
I think the `theta` variable is getting reshaped by jax to have a leading batch dimension,
and this violates the type signature.

@slinderman @patrick-kidger FYI

**Comment by murphyk** [comment]
Interestingly, this version works
```
foo = vmap(lambda y: f(theta, y))(data)
```
This avoids replicating the `theta` argument.

**Comment by murphyk** [comment]
Unfortunately the vmap+lambda trick won't work if we we return a structure that has a type decorator on its init:
```

@typechecker
class MyParams(NamedTuple):
    mean: float

theta = MyParams(2.0)

def f(params, y):
    p = MyParams(params.mean + y)
    return p

data = jnp.ones((10,2))

foo = vmap(lambda y: f(theta, y))(data)

```
I get this error
```
BeartypeCallHintParamViolation: @beartyped namedtuple_MyParams.MyParams.__new__() parameter 
mean="Traced<ShapedArray(float32[2])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([[3., 3.... violates type hint <class 'float'>,
 as <protocol "jax.interpreters.batching.BatchTracer"> 
"Traced<ShapedArray(float32[2])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([[3., 3.... not instance of float.
```

**Comment by patrick-kidger** [comment]
Ah. I think this is technically a JAX error actually.

So what's going on here is that named tuples are pytrees. Much of the pytree manipulation done internally by JAX actually involves replacing the pytree's leaves with sentinel values, e.g. `object()`. (In this case, when you cross the `jax.vmap` boundary)

It looks like this replacement is happening by going via the `__init__` method, and thus violating the typechecker.

The fix should be to get to JAX to unflatten namedtuples without calling the `__init__` method directly. (Somehow? Tuples are builtin types so Idk if you can initialise them some other way. Typically for pytrees we try to unflatten them by assigning attributes using `setattr`, to avoid precisely this kind of problem.)

I'd suggest raising this as an issue over in the JAX repo. (Or probably submitting a PR; this sounds like it'll probably be treated as low-priority.)

In the mean time you could work around this by using some other pytree type as your container. (E.g. an Equinox module, or a dataclass that's been registered via `register_pytree_node`)

**Comment by murphyk** [comment]
Thanks, Patrick. That makes sense. We may try replacing NamedTuple with dataclass (or maybe equinox, as you suggested). I'll keep you posted.

@slinderman

**Comment by leycec** [comment]
**...the JAX storm continues.** Brace yourselves. I am floundering in the dark, the seas are rising, and I have no paddle.

> The fix should be to get to JAX to unflatten `namedtuples` without calling the `__init__` method directly. (Somehow? Tuples are builtin types so Idk if you can initialise them some other way. Typically for pytrees we try to unflatten them by assigning attributes using `setattr`, to avoid precisely this kind of problem.)

`</gulp>`

I see bombs bursting in mid-air there. `typing.NamedTuple` defers to `collections.namedtuple`, which dynamically synthesizes and returns a new subclass of the builtin `tuple` type, which is immutable and thus prohibits modification – including `setattr()` calls. Indeed, `tuple` (and thus `namedtuple`) instances are initialized by the `__new__()` rather than `__init__()` dunder method. It is literally impossible to avoid calling `__new__()`.

That said, `collections.namedtuple` and thus `typing.NamedTuple` is *basically* deprecated in 2022. `collections.namedtuple` is poorly maintained, poorly documented, suffers an interminable litany of edge cases, and routinely blows up explosively. Everybody should be using the standard `@dataclasses.dataclass` decorator instead, which @beartype also fully supports but which suffers none of the issues that `collections.namedtuple` suffers. Indeed, you can perfectly mimic the feature set of `typing.NamedTuple` by just passing `frozen=True` to `@dataclasses.dataclass`: e.g.,

```python
from beartype import beartype
from dataclasses import dataclass

# Order is significant here. @beartype *MUST* be listed first in
# this decorator chain. Nothing is easy on Planet Earth. Gah!
@beartype
@dataclass(frozen=True)
class MyParams(NamedTuple):
    mean: float
```

There's still no guarantee that JAX **+** @beartype **+** `@dataclasses.dataclass` with behave sane. But at least JAX devs will (hopefully) be more attentive to integration issues involving the considerably saner `@dataclasses.dataclass` decorator.

## tl;dr

`@dataclass(frozen=True)`. Thus spake Bearathustra. :bear:

**Comment by murphyk** [comment]
Unfortunatelt just replacing NamedTuple with dataclass does not seem sufficient.
The code below gives the error
```
Traceback (most recent call last):
  File "bear.py", line 28, in <module>
    foo = vmap(lambda y: f(theta, y))(data)
TypeError: Output from batched function
 MyParams(mean=Traced<ShapedArray(float32[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([3., 3., 3., 3., 3., 3., 3., 3., 3., 3.], dtype=float32)
  batch_dim = 0) with type <class '__main__.MyParams'> is not a valid JAX type
```

```
import jax
from jax import numpy as jnp
from jax import vmap
import jax.random as jr

from jaxtyping import Array, Float, PyTree, Bool, Int, Num, jaxtyped
from typing_extensions import Protocol
from typing import Optional, Union
from dataclasses import dataclass
from beartype import beartype as typechecker
#from typeguard import typechecked as typechecker

Scalar = Union[float, Float[Array, ""]] # python float or scalar jax device array with dtype float

@typechecker
@dataclass(frozen=True)
class MyParams:
    mean: Scalar

theta = MyParams(2.0)

def f(params, y):
    p = MyParams(params.mean + y)
    return p

data = jnp.ones(10)

foo = vmap(lambda y: f(theta, y))(data)
```

**Comment by patrick-kidger** [comment]
That's because, as the error suggests, dataclasses aren't JAX types. You should register it as PyTree. (Or use `equinox.Module`, which is basically `@dataclass` + pytree registration + a few edge-case improvements all wrapped together.)

FWIW this definitely has nothing to do with `beartype`, so to spare @leycec I suggest we discuss this elsewhere? :)

**Comment by leycec** [comment]
**[Equinox](https://github.com/patrick-kidger/equinox).** Yeah, that's hot:

```python
import equinox as eqx
import jax

class Linear(eqx.Module):
    weight: jax.numpy.ndarray
    bias: jax.numpy.ndarray

    def __init__(self, in_size, out_size, key):
        wkey, bkey = jax.random.split(key)
        self.weight = jax.random.normal(wkey, (out_size, in_size))
        self.bias = jax.random.normal(bkey, (out_size,))

    def __call__(self, x):
        return self.weight @ x + self.bias
```

No surprise. That's @patrick-kidger for you. It's black magic all the way down. :smile: 

Let's pretend @beartype plays nicely with Equinox and quietly close this out. If there's anything further @beartype can do for anyone here, please don't hesitate to open up an oppressive flood of new issues.

**Comment by murphyk** [comment]
Yes, equinox does indeed solve this problem! :)
Here is a demonstration of this, for the record (since the benefits of equinox may not be obvious to others :)

```

@jaxtyped
@typechecker
#@dataclass(frozen=True) # bad
class MyParams(eqx.Module): # good
    mean: Scalar
    mean2: Float[Array, "ndim"]

theta = MyParams(1.0, jnp.ones(2))

def f(params, y):
    p = MyParams(params.mean + y, y*jnp.ones(3))
    return p

data = jnp.arange(5)

foo = vmap(lambda y: f(theta, y))(data)
print(foo)
print('mean ', foo.mean)
print('mean2 ', foo.mean2)
```
This type checks and produces the desired output:
```
MyParams(mean=f32[5], mean2=f32[5,3])
mean  [1. 2. 3. 4. 5.]
mean2
  [[0. 0. 0.]
 [1. 1. 1.]
 [2. 2. 2.]
 [3. 3. 3.]
 [4. 4. 4.]]
```

**Comment by leycec** [comment]
**Fantabulous!** Superb example as well, @murphyk. That is indeed quite non-obvious. Let's loudly reopen this as a new documentation task. We'll append your beautiful example with a bit of explanation to our existing [FAQ entry on JAX typing](https://github.com/beartype/beartype#jax-arrays).

**Comment by leycec** [comment]
**Resolved by @beartype 0.17.0,** which officially supports Equinox. Or is it that... *Equinox officially supports @beartype!?* Actually, I'm pretty sure this is a bijective relation. @leycec and @patrick-kidger both slaved over mutual keyboards to make this magic.

The future of ML now rests in your hands, capable user. @beartype and Equinox salute you. :bear: **+** :robot: = :wink:
